storey/sources.py

```diff
@@ -622,49 +620,85 @@ def __init__(
                 elif df.index.names[0] is not None:
                     body[df.index.names[0]] = index
-
-                key = None
-                if self._key_field:
-                    if isinstance(self._key_field, list):
-                        key = []
-                        for key_field in self._key_field:
-                            if key_field not in body or pandas.isna(body[key_field]):
-                                create_event = False
-                                break
-                            key.append(body[key_field])
-                    else:
-                        key = body[self._key_field]
-                        if key is None:
-                            create_event = False
-                if create_event:
+                try:
+                    key = self._get_key(body=body)
                     if self._id_field:
-                        id = body[self._id_field]
+                        line_id = self._get_by_field_or_index(
+                            field=self._id_field, body=body, field_type="id", raise_exception=False
+                        )
                     else:
-                        id = self._get_uuid()
-                    event = Event(body, key=key, id=id)
+                        line_id = self._get_uuid()
+                    element = self._get_element(body=body)
+                    event = Event(element, key=key, id=line_id)
                     await self._do_downstream(event)
-                else:
+                except self.NoneKeyException as key_error:
                     if self.context:
-                        self.context.logger.error(f"For {body} value of key {key_field} is None")
+                        self.context.logger.error(str(key_error))
+
         return await self._do_downstream(_termination_obj)
 
+    def _get_by_field_or_index(self, field, body: OrderedDict, field_type: str, raise_exception=False):
+        result = body[field]
+        if raise_exception:
+            self._is_nan_validator(result=result, body=body, field_type=field_type, field=field)
+        return result
+
+    def _field_validator(self, df, key_field, id_field, path=""):
+        path_message = get_path_message(path=path)
+        df = df.reset_index()
+        if key_field:
+            key_field = [key_field] if not isinstance(key_field, list) else key_field
+            missing_keys = list(set(key_field) - set(df.columns))
+            if missing_keys:
+                raise ValueError(f"keys {missing_keys} missing from dataframe.{path_message}")
+
+        if id_field and id_field not in df.columns:
+            raise ValueError(f"id field '{id_field}' missing from dataframe.{path_message}")
+
+    class NoneKeyException(Exception):
+        pass
 
-class CSVSource(_IterableSource, WithUUID):
+    def _is_nan_validator(self, result, body, field_type, field):
+        if pandas.isna(result) or result is None:
```

Comment:
`pandas.isna(None)` returns `True`, so the second check is redundant.
```suggestion
        if pandas.isna(result):
```

Reply:
you are right fixed